snapcraft/_legacy_loader.py

```diff
@@ -0,0 +1,84 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2020 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import importlib
+import importlib.abc
+import importlib.machinery
+import importlib.util
+import warnings
+import sys
+
+
+_VALID_V1_PLUGINS = [
+    "ant",
+    "catkin-tools",
+    "conda",
+    "dump",
+    "gradle",
+    "make",
+    "nil",
+    "python",
+    "rust",
+    "autotools",
+    "cmake",
+    "go",
+    "kbuild",
+    "maven",
+    "nodejs",
+    "qmake",
+    "scons",
+    "catkin",
+    "colcon",
+    "crystal",
+    "dotnet",
+    "godeps",
+    "kernel",
+    "meson",
+    "plainbox-provider",
+    "ruby",
+    "waf",
+]
+
+
+class LegacyPluginLoader(importlib.abc.Loader):
+    @classmethod
+    def create_module(cls, spec):
+        # Load the plugin from the new location.
+        plugin_name = spec.name.split(".")[-1]
+        return importlib.import_module(f"snapcraft.plugins.v1.{plugin_name}")
+
+    @classmethod
+    def exec_module(cls, module):
+        # Rewrite the module __name__ to have that of the legacy import path.
+        plugin_name = module.__name__.split(".")[-1]
+        module.__name__ = f"snapcraft.plugins.{plugin_name}"
+        return module
+
+
+class LegacyPluginPathFinder(importlib.machinery.PathFinder):
+    @classmethod
+    def find_spec(cls, fullname, path=None, target=None):
```

Comment:
types?

Reply:
I welcome you to try, there seems to be missing mypy/typeshed bits for this to work correctly, as soon as I add any type info I get:

```
snapcraft/_legacy_loader.py:74: error: Signature of "find_spec" incompatible with supertype "MetaPathFinder"
```
Which is correct, I implemented the more lenient `PathFinder`.

``
snapcraft/_legacy_loader.py:79: error: Argument 2 to "ModuleSpec" has incompatible type "Type[LegacyPluginLoader]"; expected "Optional[Loader]"
```
Well, yeah